### ISSUE TITLE
CGameEditorPluginMap::Items is Empty in Nod Explorer and from Scripts

### ISSUE THREAD

**Issue by skybaks** [openplanet-nl/openplanet-issues#219]
In the Nod Explorer under: "Editor->PluginMapType" there are several arrays of objects which are listed as "nodpool". All these arrays are always displayed as empty when observing them from both Nod Explorer and via scripts.

The one I am specifically interested in is `Items`. When accessing this from the maniascript editor plugin API I can see that there are objects in this array, but they still dont show up in openplanet.

Is this due to some limitation of how the game is hooked or something else? If its possible to fix this I would love it but if not I'm curious as to why these nodpools are different than other arrays.

**Comment by codecat** [comment]
It's likely related to the timing at which the Maniascript is executed versus the timing of Openplanet's Nod Explorer.

I've done some work before to make stuff like this actually possible, but I haven't made it a stable feature just yet. Maybe I'll set it up as a beta thing soon.

Regardless, I'll have to take a look at this specific case.

**Comment by XertroV** [comment]
I think this is not related to execution time. Basically, if you run `CheckItemsNodPool` in a coroutine or something, **while the editor is open with at least one item**: it seems to be able to access the items. So I think this is probably something to do with the way openplanet reads these structures.

I noticed that the NodPool structure seems to be embedded in the PluginMapType object (not a pointer). 
Inside it, there's something that looks like a pointer to a buffer with the expected number and type of `.Items`. Except the 2 uint32s following the pointer are 0x0 and then the count. So that's odd but seems to match atm at least.

```angelscript
void CheckItemsNodPool() {
    auto editor = cast<CGameCtnEditorFree>(GetApp().Editor);
    trace('Nb PluginMapType.Items: ' + editor.PluginMapType.Items.Length);
    auto pmt = cast<CSmEditorPluginMapType>(editor.PluginMapType);
    trace('Nb pmt.Items: ' + pmt.Items.Length);
    auto editorTy = Reflection::GetType("CGameCtnEditorFree");
    auto pmtMember = editorTy.GetMember("PluginMapType");
    auto pmtTy = Reflection::GetType("CSmEditorPluginMapType");
    auto itemsMember = pmtTy.GetMember("Items");
    if (itemsMember.Offset < 0xffff) {
        auto pmtPointer = Dev::GetOffsetUint64(editor, pmtMember.Offset);
        auto nodPoolPtr = pmtPointer + itemsMember.Offset;
        trace('pmtPointer: ' + pmtPointer);
        trace('nodPoolPtr: ' + nodPoolPtr);
        IO::SetClipboard('' + nodPoolPtr);
        auto np2 = nodPoolPtr + 0x20;
        auto count = Dev::ReadUInt32(nodPoolPtr + 0x2c);
        trace('count: ' + count);
        auto nodPool = Dev::GetOffsetNod(pmt, itemsMember.Offset + 0x20);
        auto nod1 = Dev::GetOffsetNod(nodPool, 0x0);
        auto nodPtr = Dev::GetOffsetUint64(nodPool, 0x0);
        trace('nodPtr: ' + nodPtr);
        if (nod1 !is null) {
            auto nodTy = Reflection::TypeOf(nod1);
            if (nodTy !is null)
                trace('nod is of type; ' + nodTy.Name);
        }
        trace('if no type, nod type wasn\'t available');
    }
}
```

----------

Additionally: During ML execution, the number of `.Items` is still 0 from angelscript. 

I checked using this editor plugin script + the ML Event Inspector + Some plugin code. 

```maniascript
#RequireContext CMapEditorPlugin

declare Text ML_Page_Src = """
<manialink version="3" name="TestAsCbInEditor">
<script> <!--
main() {
 while(True)
 {
	yield;
	SendCustomEvent("MLHook_Event_Editor_Angelscript_Cb", [""^Editor.Items.count]);
 }
}
--> </script>
</manialink>
""";

declare layer <=> UILayerCreate();
layer.ManialinkPage = ML_Page_Src;

while (True) {
  yield;
  // no send custom event, this is useful for debugging via event inspector
  // LayerCustomEvent(layer, "MLHook_Event_Editor_Layer", [""^Items.count, ""^layer.LocalPage]); // ""^This
}
```

plugin code:

```angelscript
#if DEPENDENCY_MLHOOK

uint count = 0;

// inject via `MLHook::RegisterMLHook(EditorExecHook(), "Editor_Angelscript_Cb");`
class EditorExecHook : MLHook::HookMLEventsByType {
    EditorExecHook() {
        super("Editor_Angelscript_Cb");
    }

    void OnEvent(MLHook::PendingEvent@ event) override {
        // don't need the event, just want to run at ML time
        if (count < 10) {
            CheckItemsNodPool();
        }
        count++;
    }
}

#endif
```

**Comment by codecat** [comment]
You're correct that the items exist within the buffer, however the buffer length is always reported as 0, and that **must be respected**. You're looking at the buffer size itself (offset `0x2C`), which is just how big the buffer is. If you add an item to the buffer and then remove it again, the buffer length is reduced to 0, but the allocated size will still be 1.

The buffer you are using at `0x20` is the linear buffer, which I've been told will be removed later as it's currently only there for compatibility reasons. This could also be a reason why the length is 0 in scripts, as that's what we're still currently using to access nod pools, but I'm not 100% sure about that. It's not normally 0 anyway.

The reason the items seem to remain valid after removing them is due to this being a memory pool - the objects remain in memory and then just get re-used later. Only 1 allocation and free will be performed.

To conclude this comment, yes, I need to implement the **proper** way instead of using the compatibility values, and will need to do some more research to figure out how exactly that should work.

**Comment by codecat** [comment]
Did some more digging today, and I can't get anything useful out of the nodpool's raw block data either, all elements in the memory blocks are listed as "free" (eg: memory is unused, new elements can re-use this memory). Accessing these objects would be unsafe.

![image](https://user-images.githubusercontent.com/136534/234018694-ecb97a64-cff4-4218-9ab1-40803ae97096.png)

It looks line the linear buffer is not "bugged" or "incorrect" (at least on Openplanet's site), this in fact seems like the intended design of the `Items` property.

As the "first free" handle increases each frame (the "version" portion of the handle), I believe this memory is being overridden each time and there is still some kind of timing issue at play. I assume it's accessible during Maniascript execution in the map type context.

**Comment by XertroV** [comment]
I had another look out of curiosity and found some things:

- The buffer at 0x20 is indeed populated and cleared every frame. 
- It appears that the ScriptAnchoredItems in those blocks have pointers in them set and cleared each frame. There's some code at `Trackmania.exe+117F642 - 48 89 41 18           - mov [rcx+18],rax` that you can nop to see the pointer(s) that were populated in the script anchored object. (doesn't seem to break anything).

I think something must have been wrong with my code before -- it does indeed seem to be populated ~~during ML exec~~ and cleared once per frame. (I didn't confirm that it was populated during ML exec)

**Comment by XertroV** [comment]
So I realized that my mlhook thing was never going to work b/c those kind of hooks aren't run during ML execution. (derp)

However, I did set up an intercept for SendCustomEvent and even then the buffer wasn't populated. This is confusing b/c it's triggered by:
```
	SendCustomEvent("MLHook_Event_Editor_Angelscript_Cb", [""^Editor.Items.count, ""^Editor.AnchorData.count]);
	SendCustomEvent("test", []);
```

So I don't get how in ML it can be populated, then in anglescript it's not populated, when i'm directly intercepting that call and reading the raw memory values. Esp when it definitely is populated at some point.

Unless `Editor.Items` in the ML is a different PluginMapType instance? (But I only find one scanning memory)

**Comment by codecat** [comment]
What is the output of that custom event?

**Comment by XertroV** [comment]
> What is the output of that custom event?

Do you mean the return value of `SendCustomEvent`? that's `Void` i think.

Otherwise it doesn't do anything, just passed through the event pipeline and ignored since the event type doesn't match anything. 

`MLHook_*` events are blocked by MLHook, though.

**Comment by codecat** [comment]
I mean the array you are passing to the event, what does that say the item count is?

**Comment by XertroV** [comment]
> I mean the array you are passing to the event, what does that say the item count is?

Oh, it reads the correct number of items.
(Briefly scanning above, it looks like I might have missed this detail -- oops!)

This matches what I've seen in-memory via a strategy of create-long-frame-times-in-ML + lots of items. 
I created like 6k items and then added a for loop that counted up to 200k each time. (1m created a 'this ML takes too long' exception).
The changes in memory were much more obvious this way. It was clear that the linear buffer was being populated and cleared each frame. (along with nodpool blocks being written and cleared)

**Comment by bmx22c** [comment]
Any news on that topic ? 👀

**Comment by XertroV** [comment]
> Any news on that topic ? 👀

Yes, got it working via MLHook. This also confirms that it is indeed an execution timing thing.

The next version of MLHook will add support for running code during ML exec so you will be able to access `.Items`.

**Comment by bmx22c** [comment]
Hell yeah, you rock! Thanks a lot!
